File: src/libmodule.js
// Copyright 2021 Roy T. Hashimoto. All Rights Reserved.
// @ts-ignore
const mod_methods = {
  $mod_method_support__postset: 'mod_method_support();',
  $mod_method_support: function() {
    const hasAsyncify = typeof Asyncify === 'object';

    const mapIdToModule = new Map();
    const mapVTabToModule = new Map();
    const mapCursorToModule = new Map();

    const closedVTabs = hasAsyncify ? new Set() : null;
    const closedCursors = hasAsyncify ? new Set() : null;

    class Value {
      constructor(ptr, type) {
        this.ptr = ptr;
        this.type = type;
      }

      set(v) {
        switch (this.type) {
          case 's':
            const length = lengthBytesUTF8(v);
            const p = ccall('sqlite3_malloc', 'number', ['number'], [length + 1]);
            stringToUTF8(v, p, length + 1);
            setValue(this.ptr, p, 'i32');
            break;
          default:
            setValue(this.ptr, v, this.type);
            break;
        }
      }
    }

    /** Field offsets within SQLite C structs.
     *  @type {Map<string, { size: number, offsets: Array<number>}>}
     */
    const mapStructToLayout = new Map();
    _modStruct = function(zName, iSize, nFields, pOffsets) {
      mapStructToLayout.set(UTF8ToString(zName), {
        size: iSize,
        offsets: Array.from(new Uint32Array(HEAP8.buffer, pOffsets, nFields))
      });
    };

    function unpack_sqlite3_index_info(p) {
      const layout = mapStructToLayout.get('sqlite3_index_info');
      const offset = layout.offsets;
      const struct = {};
      struct['nConstraint'] = getValue(p + offset[0], 'i32');
      struct['aConstraint'] = [];
      const constraintPtr = getValue(p + offset[1], 'i32');
      const constraintSize = mapStructToLayout.get('sqlite3_index_constraint').size;
      for (let i = 0; i < struct['nConstraint']; ++i) {
        struct['aConstraint'].push(
          unpack_sqlite3_index_constraint(constraintPtr + i * constraintSize));
      }
      struct['nOrderBy'] = getValue(p + offset[2], 'i32');
      struct['aOrderBy'] = [];
      const orderPtr = getValue(p + offset[3], 'i32');
      const orderSize = mapStructToLayout.get('sqlite3_index_orderby').size;
      for (let i = 0; i < struct['nOrderBy']; ++i) {
        struct['aOrderBy'].push(
          unpack_sqlite3_index_orderby(orderPtr + i * orderSize));
      }
      
      struct['aConstraintUsage'] = [];
      for (let i = 0; i < struct['nConstraint']; ++i) {
        struct['aConstraintUsage'].push({
          'argvIndex': 0,
          'omit': false
        });
      }
      struct['idxNum'] = getValue(p + offset[5], 'i32');
      struct['idxStr'] = null;
      struct['orderByConsumed'] = !!getValue(p + offset[8], 'i8');
      struct['estimatedCost'] = getValue(p + offset[9], 'double');
      struct['estimatedRows'] = getValue(p + offset[10], 'i64');
      struct['idxFlags'] = getValue(p + offset[11], 'i32');
      struct['colUsed'] = getValue(p + offset[12], 'i64');
      return struct;
    }

    function unpack_sqlite3_index_constraint(p) {
      const layout = mapStructToLayout.get('sqlite3_index_constraint');
      const offset = layout.offsets;
      const struct = {};
      struct['iColumn'] = getValue(p + offset[0], 'i32');
      struct['op'] = getValue(p + offset[1], 'i8');
      struct['usable'] = !!getValue(p + offset[2], 'i8');
      return struct;
    }

    function unpack_sqlite3_index_orderby(p) {
      const layout = mapStructToLayout.get('sqlite3_index_orderby');
      const offset = layout.offsets;
      const struct = {};
      struct['iColumn'] = getValue(p + offset[0], 'i32');
      struct['desc'] = !!getValue(p + offset[1], 'i8');
      return struct;
    }

    function pack_sqlite3_index_info(p, struct) {
      const layout = mapStructToLayout.get('sqlite3_index_info');
      const offset = layout.offsets;
      const usagePtr = getValue(p + offset[4], 'i32');
      const usageSize = mapStructToLayout.get('sqlite3_index_constraint_usage').size;
      for (let i = 0; i < struct['nConstraint']; ++i) {
        pack_sqlite_index_constraint_usage(
          usagePtr + i * usageSize,
          struct['aConstraintUsage'][i]);
      }
      setValue(p + offset[5], struct['idxNum'], 'i32');
      if (typeof struct['idxStr'] === 'string') {
        const length = lengthBytesUTF8(struct['idxStr']);
        const z = ccall('sqlite3_malloc', 'number', ['number'], [length + 1]);
        stringToUTF8(struct['idxStr'], z, length + 1);
        setValue(p + offset[6], z, 'i32');
        setValue(p + offset[7], 1, 'i32');
      }
      setValue(p + offset[8], struct['orderByConsumed'], 'i32');
      setValue(p + offset[9], struct['estimatedCost'], 'double');
      setValue(p + offset[10], struct['estimatedRows'], 'i64');
      setValue(p + offset[11], struct['idxFlags'], 'i32');
    }

    function pack_sqlite_index_constraint_usage(p, struct) {
      const layout = mapStructToLayout.get('sqlite3_index_constraint_usage');
      const offset = layout.offsets;
      setValue(p + offset[0], struct['argvIndex'], 'i32');
      setValue(p + offset[1], struct['omit'] ? 1 : 0, 'i8');
    }

    Module['createModule'] = function(db, zName, module, appData) {
      if (hasAsyncify) {
        // Inject Asyncify method.
        module['handleAsync'] = Asyncify.handleAsync;
      }

      const key = mapIdToModule.size;
      mapIdToModule.set(key, {
        module: module,
        appData: appData
      });

      // Set bits for the provided functions.
      let flags = 0;
      if (module['xCreate']) flags |= 1 << 0;
      if (module['xConnect']) flags |= 1 << 1;
      if (module['xBestIndex']) flags |= 1 << 2;
      if (module['xDisconnect']) flags |= 1 << 3;
      if (module['xDestroy']) flags |= 1 << 4;
      if (module['xOpen']) flags |= 1 << 5;
      if (module['xClose']) flags |= 1 << 6;
      if (module['xFilter']) flags |= 1 << 7;
      if (module['xNext']) flags |= 1 << 8;
      if (module['xEof']) flags |= 1 << 9;
      if (module['xColumn']) flags |= 1 << 10;
      if (module['xRowid']) flags |= 1 << 11;
      if (module['xUpdate']) flags |= 1 << 12;
      if (module['xBegin']) flags |= 1 << 13;
      if (module['xSync']) flags |= 1 << 14;
      if (module['xCommit']) flags |= 1 << 15;
      if (module['xRollback']) flags |= 1 << 16;
      if (module['xFindFunction']) flags |= 1 << 17;
      if (module['xRename']) flags |= 1 << 18;

      return ccall(
        'create_module',
        'number',
        ['number', 'string', 'number', 'number'],
        [db, zName, key, flags]);
    };

    _modCreate = function(db, pModuleId, argc, argv, pVTab, pzErr) {
      const m = mapIdToModule.get(pModuleId);
      mapVTabToModule.set(pVTab, m);
      if (hasAsyncify) {
        closedVTabs.delete(pVTab);
        for (const vTab of closedVTabs) {
          mapVTabToModule.delete(vTab);
        }
      }
      argv = Array.from(new Uint32Array(HEAP8.buffer, argv, argc))
        .map(p => UTF8ToString(p));
      return m.module['xCreate'](db, m.appData, argv, pVTab, new Value(pzErr, 's'));
    };

    _modConnect = function(db, pModuleId, argc, argv, pVTab, pzErr) {
      const m = mapIdToModule.get(pModuleId);
      mapVTabToModule.set(pVTab, m);
      if (hasAsyncify) {
        closedVTabs.delete(pVTab);
        for (const vTab of closedVTabs) {
          mapVTabToModule.delete(vTab);
        }
      }
      argv = Array.from(new Uint32Array(HEAP8.buffer, argv, argc))
        .map(p => UTF8ToString(p));
      return m.module['xConnect'](db, m.appData, argv, pVTab, new Value(pzErr, 's'));
    };

    _modBestIndex = function(pVTab, pIndexInfo) {
      const m = mapVTabToModule.get(pVTab);
      const indexInfo = unpack_sqlite3_index_info(pIndexInfo);
      const result = m.module['xBestIndex'](pVTab, indexInfo);
      pack_sqlite3_index_info(pIndexInfo, indexInfo);
      return result;
    };

    _modDisconnect = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      if (hasAsyncify) {
        closedVTabs.add(pVTab);
      } else {
        mapVTabToModule.delete(pVTab);
      }
      return m.module['xDisconnect'](pVTab);
    };

    _modDestroy = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      if (hasAsyncify) {
        closedVTabs.add(pVTab);
      } else {
        mapVTabToModule.delete(pVTab);
      }
      return m.module['xDestroy'](pVTab);
    };

    _modOpen = function(pVTab, pCursor) {
      const m = mapVTabToModule.get(pVTab);
      mapCursorToModule.set(pCursor, m);
      if (hasAsyncify) {
        closedCursors.delete(pCursor);
        for (const cursor of closedCursors) {
          mapCursorToModule.delete(cursor);
        }
      }
      return m.module['xOpen'](pVTab, pCursor);
    };

    _modClose = function(pCursor) {
      const m = mapCursorToModule.get(pCursor);
      if (hasAsyncify) {
        closedCursors.add(pCursor);
      } else {
        mapCursorToModule.delete(pCursor);
      }
      return m.module['xClose'](pCursor);
    };

    _modEof = function(pCursor) {
      const m = mapCursorToModule.get(pCursor);
      return m.module['xEof'](pCursor) ? 1 : 0;
    };

    _modFilter = function(pCursor, idxNum, idxStr, argc, argv) {
      const m = mapCursorToModule.get(pCursor);
      idxStr = idxStr ? UTF8ToString(idxStr) : null;
      argv = new Uint32Array(HEAP8.buffer, argv, argc);
      return m.module['xFilter'](pCursor, idxNum, idxStr, argv);
    };

    _modNext = function(pCursor) {
      const m = mapCursorToModule.get(pCursor);
      return m.module['xNext'](pCursor);
    };

    _modColumn = function(pCursor, pContext, iCol) {
      const m = mapCursorToModule.get(pCursor);
      return m.module['xColumn'](pCursor, pContext, iCol);
    };

    _modRowid = function(pCursor, pRowid) {
      const m = mapCursorToModule.get(pCursor);
      return m.module['xRowid'](pCursor, new Value(pRowid, 'i64'));
    };

    _modUpdate = function(pVTab, argc, argv, pRowid) {
      const m = mapVTabToModule.get(pVTab);
      argv = new Uint32Array(HEAP8.buffer, argv, argc);
      return m.module['xUpdate'](pVTab, argv, new Value(pRowid, 'i64'));
    };

    _modBegin = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      return m.module['xBegin'](pVTab);
    };

    _modSync = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      return m.module['xSync'](pVTab);
    };

    _modCommit = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      return m.module['xCommit'](pVTab);
    };

    _modRollback = function(pVTab) {
      const m = mapVTabToModule.get(pVTab);
      return m.module['xRollback'](pVTab);
    };

    _modRename = function(pVTab, zNew) {
      const m = mapVTabToModule.get(pVTab);
      zNew = UTF8ToString(zNew);
      return m.module['xRename'](pVTab, zNew);
    }
  }
};

// @ts-ignore
const MOD_METHOD_NAMES = [
  "modStruct",
  "modCreate",
  "modConnect",
  "modBestIndex",
  "modDisconnect",
  "modDestroy",
  "modOpen",
  "modClose",
  "modFilter",
  "modNext",
  "modEof",
  "modColumn",
  "modRowid",

  // Optional methods.
  "modUpdate",
  "modBegin",
  "modSync",
  "modCommit",
  "modRollback",
  "modFindFunction",
  "modRename",
];
for (const method of MOD_METHOD_NAMES) {
  mod_methods[method] = function() {};
  mod_methods[`${method}__deps`] = ['$mod_method_support'];
}
mergeInto(LibraryManager.library, mod_methods);

File: src/libmodule.c
// Copyright 2021 Roy T. Hashimoto. All Rights Reserved.
#include <stddef.h>
#include <string.h>
#include <emscripten.h>
#include <sqlite3.h>

extern int modStruct(const char* zName, int iSize, int nFields, int* pOffsets);

extern int modCreate(sqlite3*, void *pAux,
               int argc, const char *const*argv,
               sqlite3_vtab* pVTab, char**);
extern int modConnect(sqlite3*, void *pAux,
               int argc, const char *const*argv,
               sqlite3_vtab* pVTab, char**);
extern int modBestIndex(sqlite3_vtab *pVTab, sqlite3_index_info*);
extern int modDisconnect(sqlite3_vtab *pVTab);
extern int modDestroy(sqlite3_vtab *pVTab);
extern int modOpen(sqlite3_vtab *pVTab, sqlite3_vtab_cursor *pCursor);
extern int modClose(sqlite3_vtab_cursor*);
extern int modFilter(sqlite3_vtab_cursor*, int idxNum, const char *idxStr,
                int argc, sqlite3_value **argv);
extern int modNext(sqlite3_vtab_cursor*);
extern int modEof(sqlite3_vtab_cursor*);
extern int modColumn(sqlite3_vtab_cursor*, sqlite3_context*, int);
extern int modRowid(sqlite3_vtab_cursor*, sqlite3_int64 *pRowid);
extern int modUpdate(sqlite3_vtab *, int, sqlite3_value **, sqlite3_int64 *);
extern int modBegin(sqlite3_vtab *pVTab);
extern int modSync(sqlite3_vtab *pVTab);
extern int modCommit(sqlite3_vtab *pVTab);
extern int modRollback(sqlite3_vtab *pVTab);
// extern int modFindFunction(sqlite3_vtab *pVtab, int nArg, const char *zName,
//                        void (**pxFunc)(sqlite3_context*,int,sqlite3_value**),
//                        void **ppArg);
extern int modRename(sqlite3_vtab *pVtab, const char *zNew);

static int xCreate(
  sqlite3* db,
  void *pAux,
  int argc,
  const char *const*argv,
  sqlite3_vtab **ppVTab,
  char** pzErr) {
  *ppVTab = (sqlite3_vtab*)sqlite3_malloc(sizeof(sqlite3_vtab));
  const int result = modCreate(db, pAux, argc, argv, *ppVTab, pzErr);
  if (result != SQLITE_OK) {
    sqlite3_free(*ppVTab);
    *ppVTab = 0;
  }
  return result;
}

static int xConnect(
  sqlite3* db,
  void *pAux,
  int argc,
  const char *const*argv,
  sqlite3_vtab **ppVTab,
  char** pzErr) {
  *ppVTab = (sqlite3_vtab*)sqlite3_malloc(sizeof(sqlite3_vtab));
  const int result = modConnect(db, pAux, argc, argv, *ppVTab, pzErr);
  if (result != SQLITE_OK) {
    sqlite3_free(*ppVTab);
    *ppVTab = 0;
  }
  return result;
}

static int xOpen(sqlite3_vtab *pVTab, sqlite3_vtab_cursor **ppCursor) {
  *ppCursor = (sqlite3_vtab_cursor*)sqlite3_malloc(sizeof(sqlite3_vtab_cursor));
  return modOpen(pVTab, *ppCursor);
}

static void module_layout() {
#define LAYOUT_BEGIN(TYPE) \
  static int offsets_##TYPE[] = {
#define LAYOUT_DECLARE(TYPE, MEMBER) \
    offsetof(struct TYPE, MEMBER),
#define LAYOUT_END(TYPE) \
  }; \
  modStruct( \
    #TYPE, \
    sizeof(struct TYPE), \
    sizeof(offsets_##TYPE) / sizeof(int), \
    offsets_##TYPE);

  LAYOUT_BEGIN(sqlite3_index_info)
  LAYOUT_DECLARE(sqlite3_index_info, nConstraint)
  LAYOUT_DECLARE(sqlite3_index_info, aConstraint)
  LAYOUT_DECLARE(sqlite3_index_info, nOrderBy)
  LAYOUT_DECLARE(sqlite3_index_info, aOrderBy)
  LAYOUT_DECLARE(sqlite3_index_info, aConstraintUsage)
  LAYOUT_DECLARE(sqlite3_index_info, idxNum)
  LAYOUT_DECLARE(sqlite3_index_info, idxStr)
  LAYOUT_DECLARE(sqlite3_index_info, needToFreeIdxStr)
  LAYOUT_DECLARE(sqlite3_index_info, orderByConsumed)
  LAYOUT_DECLARE(sqlite3_index_info, estimatedCost)
  LAYOUT_DECLARE(sqlite3_index_info, estimatedRows)
  LAYOUT_DECLARE(sqlite3_index_info, idxFlags)
  LAYOUT_DECLARE(sqlite3_index_info, colUsed)
  LAYOUT_END(sqlite3_index_info)

  LAYOUT_BEGIN(sqlite3_index_constraint)
  LAYOUT_DECLARE(sqlite3_index_constraint, iColumn)
  LAYOUT_DECLARE(sqlite3_index_constraint, op)
  LAYOUT_DECLARE(sqlite3_index_constraint, usable)
  LAYOUT_DECLARE(sqlite3_index_constraint, iTermOffset)
  LAYOUT_END(sqlite3_index_constraint)

  LAYOUT_BEGIN(sqlite3_index_orderby)
  LAYOUT_DECLARE(sqlite3_index_orderby, iColumn)
  LAYOUT_DECLARE(sqlite3_index_orderby, desc)
  LAYOUT_END(sqlite3_index_orderby)

  LAYOUT_BEGIN(sqlite3_index_constraint_usage)
  LAYOUT_DECLARE(sqlite3_index_constraint_usage, argvIndex)
  LAYOUT_DECLARE(sqlite3_index_constraint_usage, omit)
  LAYOUT_END(sqlite3_index_constraint_usage)

#undef LAYOUT_BEGIN
#undef LAYOUT_DECLARE
#undef LAYOUT_END  
}

int EMSCRIPTEN_KEEPALIVE create_module(
  sqlite3* db,
  const char* zName,
  void* pClientData,
  int flags) {
  // Tell Javascript the layout of C structs.
  static int ready = 0;
  if (!ready) {
    module_layout();
    ready = 1;
  }

  sqlite3_module* module = (struct sqlite3_module*)sqlite3_malloc(sizeof(sqlite3_module));
  memset(module, 0, sizeof(sqlite3_module));
  module->iVersion = 1;
  if (flags & (1 << 0)) module->xCreate = xCreate;
  module->xConnect = xConnect;
  module->xBestIndex = modBestIndex;
  module->xDisconnect = modDisconnect;
  module->xDestroy = modDestroy;
  module->xOpen = xOpen;
  module->xClose = modClose;
  module->xFilter = modFilter;
  module->xNext = modNext;
  module->xEof = modEof;
  module->xColumn = modColumn;
  module->xRowid = modRowid;
  if (flags & (1 << 12)) module->xUpdate = modUpdate;
  if (flags & (1 << 13)) module->xBegin = modBegin;
  if (flags & (1 << 14)) module->xSync = modSync;
  if (flags & (1 << 15)) module->xCommit = modCommit;
  if (flags & (1 << 16)) module->xRollback = modRollback;
  // xFindFunction not supported
  if (flags & (1 << 18)) module->xRename = modRename;
  return sqlite3_create_module(db, zName, module, pClientData);
}
